pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/PlanFragmenter.java

```diff
@@ -126,49 +124,65 @@ public PlanNode visitExchange(ExchangeNode node, Context context) {
     if (!isPlanFragmentSplitter(node)) {
       return process(node, context);
     }
-    int currentPlanFragmentId = context._previousPlanFragmentId;
-    int nextPlanFragmentId = ++context._currentPlanFragmentId;
-    // Set previous PlanFragment ID in the context to be the next PlanFragment ID to be used by the child node.
-    context._previousPlanFragmentId = nextPlanFragmentId;
-    PlanNode nextPlanFragmentRoot = node.getInputs().get(0).visit(this, context);
 
+    // Split the ExchangeNode to a MailboxReceiveNode and a MailboxSendNode, where MailboxReceiveNode is the leave node
+    // of the current PlanFragment, and MailboxSendNode is the root node of the next PlanFragment.
+    int receiverPlanFragmentId = context._currentPlanFragmentId;
+    int senderPlanFragmentId = context._nextPlanFragmentId.getAndIncrement();
```

Comment:
Modified to `MutableInt` to avoid confusion.
IMO this is the correct way to handle this recursion. Basically we share `_nextPlanFragmentId` across all PlanFragments to ensure uniqueness of PlanFragment ID. Added a comment to explain this

Reply:
yes i think that's correct. i was wondering if we can make it even more straightforward. 

- IMO the "context" is pre-fragment
- but the global next ID is per planFragmenter --> we can create a global next ID tracker inside the planFragmenter instead (ditch the static INSTANCE approach)

WDYT?

Reply:
Good point, refactored again